js/ad_manager_vast.js

```diff
@@ -1875,6 +2142,664 @@ OO.Ads.manager(function(_, $) {
+          if (volume) {
+            this.sendVpaidTracking('unmute');
+          } else {
+            this.sendVpaidTracking('mute');
+          }
+        break;
+
+        case VPAID_EVENTS.AD_USER_ACCEPT_INVITATION:
+          this.sendVpaidTracking('acceptInvitation');
+        break;
+
+        case VPAID_EVENTS.AD_USER_MINIMIZE:
+          this.sendVpaidTracking('collapse');
+        break;
+
+        case VPAID_EVENTS.AD_USER_CLOSE:
+          this.sendVpaidTracking('close');
+        break;
+
+        case VPAID_EVENTS.AD_PAUSED:
+          this.sendVpaidTracking('pause');
+          fromPause = true;
+        break;
+
+        case VPAID_EVENTS.AD_PLAYING:
+          this.sendVpaidTracking('resume');
+        break;
+      };
+    }, this);
+
+    /**
+     * Set variables to its default state
+     * @private
+     * @method Vast#_resetAdState
+     */
+    var _resetAdState = _.bind(function() {
+      _removeListeners(currentAd.vpaidAd);
+      currentAd            = null;
+      this.currentAdBeingLoaded = null;
+      this.format               = null;
+      this.node                 = null;
+      adLoaded                  = false;
+    }, this);
+
+    /**
+     * Remove any new lines, line breaks and spaces from string.
+     * @private
+     * @method Vast#_cleanString
+     * @return {string} String with no spaces
+     */
+    var _cleanString = function(string) {
+      return string.replace(/\r?\n|\r/g, '').trim();
+    };
+
+    /**
+     * Check for clickthrough url
+     * @private
+     * @method Vast#_hasClickUrl
+     * @return {object} Ad to look for the clickthrough
+     */
+    var _hasClickUrl = function(ad) {
+      if (ad && ad.data) {
+        var  videoClickTracking = ad.data.videoClickTracking;
+        if (videoClickTracking.clickThrough) {
+          return videoClickTracking.clickThrough.length > 0;
+        }
+      }
+      return false;
+    };
+
+    /**
+     * Check wether or not a vpaid ad is valid by checking the version and the minimum required functions
+     * This is only required for VPAID ads
+     * @method Vast#_isValidVPaid
+     * @private
+     * @return {boolean} VPaid validated value
+     */
+    var _isValidVPaid = _.bind(function() {
+      var vpaidVersion = null;
+      try{
+        //TODO: Do we want int here? If so, consider var name vpaidMajorVersion
+        vpaidVersion = parseInt(currentAd.vpaidAd.handshakeVersion('2.0'));
+      } catch (e) {
+        OO.log("VPAID 2.0: Error while fetching VPAID 2.0 creative handshakeVersion - " + e)
+      }
+
+      var isValid = true;
+
+      if (vpaidVersion !== 2) {
+        OO.log('VPaid Ad Unit version is not supported.');
+        isValid = false;
+      }
+
+      var requiredFunctions = ['handshakeVersion', 'initVpaidAd', 'startAd', 'stopAd', 'skipAd', 'resizeAd',
+                               'pauseAd', 'resumeAd', 'expandAd', 'collapseAd', 'subscribe', 'unsubscribe'];
+      _.each(requiredFunctions, function(fn) {
```

Comment:
oh,  My bad here we should check against ad unit functions. also it should be initAd in the array.